### DISCUSSION TITLE
Protocol versioning for FF-Gateway

### DISCUSSION THREAD

**Discussion by Matthbo** [frankframework/frankframework#7068] [Ideas]
We should have a way to identify that modules/applications that communicate over the frankframework-gateway are compatible with each other, for example that the deployed version of the frank-flow works with gateway deployed by the framework.

A good way to do this is through protocol versioning.
The applications that communicate through the gateway should tell the gateway what protocol version they are using, if the gateway doesn't support a certain version, because it's too old or maybe newer than the deployed gateway can handle, then the gateway should send back a special message for those applications to handle.

Having some sort of protocol versioning solves the issue that applications aren't compatible with other applications using the gateway or the gateway itself, this also gives back a more useful error or logging that can inform the people deploying and managing these applications.

**Comment by nielsm5** [comment]
This is what serialization id's are for, Hazelcast should take care of all of this.
I'm quite curious what kind if incompatibility you're talking about.

**Reply by Matthbo** [reply]
Well it started from the fact that the Frank!Flow with its new backend isn't compatible with older versions of the framework.
I can easily see a world where an application that depends on the frank!bus could be too old or too new for a deployed version of the framework, making them incompatible to properly talk to each other for some requests and resulting in strange errors that is not user friendly.

Just like an API it is probably a smart idea to keep versions so applications using it can keep on understanding each other even if there was a breaking change.
(API example `PUT /api/v2/adapter`, `PUT /api/v3/configuration/adapter` which does the same thing but applications using API v2 are still able to work until the version where the v2 API gets disabled, maybe the frank!bus could also do something like this too keep old versions running for a short while giving people time to upgrade)
